relayer/relayer/src/handlers/azero.rs

```diff
@@ -0,0 +1,281 @@
+
+    #[error("Task join error")]
+    Join(#[from] JoinError),
+}
+
+pub struct AlephZeroEventsHandler;
+
+impl AlephZeroEventsHandler {
+    pub async fn run(
+        config: Arc<Config>,
+        eth_signed_connection: Arc<SignedEthConnection>,
+        mut azero_events_receiver: mpsc::Receiver<AzeroMostEvents>,
+        block_seal_sender: mpsc::Sender<u32>,
+        circuit_breaker_sender: broadcast::Sender<CircuitBreakerEvent>,
+        mut circuit_breaker_receiver: broadcast::Receiver<CircuitBreakerEvent>,
+    ) -> Result<CircuitBreakerEvent, AlephZeroEventsHandlerError> {
+        loop {
+            debug!("Ping");
+
+            select! {
+                cb_event = circuit_breaker_receiver.recv() => {
+                    warn!("Exiting due to a circuit breaker event {cb_event:?}");
+                    return Ok(cb_event?);
+                },
+
+                Some(azero_events) = azero_events_receiver.recv() => {
+                    let AzeroMostEvents {
+                        events,
+                        to_block,
+                        ..
+                    } = azero_events;
+
+                    info!("Received a batch of {} events", events.len());
+
+                    let config = Arc::clone(&config);
+                    let eth_signed_connection = Arc::clone(&eth_signed_connection);
+                    let block_seal_sender = block_seal_sender.clone ();
+                    let circuit_breaker_sender = circuit_breaker_sender.clone ();
+
+                    // spawn non-blocking task to handle all events w-out blocking the events publisher
+                    tokio::spawn(async move {
+                        let mut tasks = JoinSet::new();
+                        for event in events {
+                            // spawn each handler in separate task as it's time consuming
+                            tasks.spawn(AlephZeroEventHandler::handle_event(
+                                event,
+                                Arc::clone(&config),
+                                Arc::clone(&eth_signed_connection),
+                            ));
+                        }
+
+                        // wait for all concurrent handler tasks to finish
+                        info!("Awaiting all handler tasks to finish");
+
+                        while let Some(result) = tasks.join_next().await {
+                            match result? {
+                                Ok(_) => {},
+                                Err(why) => {
+                                    warn!("Event handler failed {why:?}, opening circuit breaker");
+                                    circuit_breaker_sender.send(CircuitBreakerEvent::AlephZeroEventHandlerFailure)?;
+                                },
+                            }
+                        }
+
+                        // marks the batch as done and notifies redis manager
+                        info!("Marking all events up to block {to_block} as handled");
```

Comment:
I don't think this is correct, as you spawn independent tasks per block, which means that you can complete processing of one block but something breaks in processing of the previous one a bit later - in that case redis value would be incorrect.

Reply:
- we can assume that handling each batch take approximately the same amount of time, which is not an unreasonable assumption
- if we don't want to make that assumption we can make the channel have a buffer of 1

Reply:
We definitely don't want to make that assumption - empty blocks or ones that have all the transactions already processed can be processed much faster than other blocks (and there are scenarios where the previous blocks are more time-consuming to process).

I don't see how sending single event batches help here - their processing still can reorder.

What could work here is creating a one-shot for each batch of events and then receiving from these channels  sequentially in some select along other operation (in azero listener or redis manager, not sure). Not sure how easy to implement that is.

Reply:
I have no idea how such cache invalidation would be supposed to work - would you store the initial value in cache again and restart the listener? That's pretty inefficient, to make such mechanism you probably need to solve the original problem anyway

Reply:
as discussed reverted to the version that awaits batch ack in the interest of time, we will devise something after the audit.